### PR TITLE
Avoid crashes caused by ammo conversion

### PR DESCRIPTION
**Pull request by DangerNoodle**
See issue https://github.com/CleverRaven/Cataclysm-DDA/issues/19429. Converting a magazine-fed firearm to a caliber that it lacks a magazine definition for will cause a crash when the item is examined. This pull request eliminates all known instances that will cause this problem, but does not eliminate whatever might be the underlying cause.

A few convertible magazine-fed firearms could be converted to 4.6x30mm and/or 5.7x24mm, but lacked magazines defined for that chambering.

Gave the American-180 and Calico M960 conversion data for 4.6mm. Gave the SIG Mosquito conversion data for 4.6 and 5.7 mm.

Additionally implemented fixes for Extended Realistic Weapons. Conversion to .410 bore for magazine-fed firearms is now facilitated by adding the Saiga-410 shotgun, with requisite magazines. To reduce the massive amount of edits that would be required, the .410 mod now can only be used on 12 gauge firearms, thus making the Saiga-12 the only weapon to be overidden to support this change.

Additionally fixed an oversight wherein the .410 conversion kit was not in any itemgroup.

### CODE REVIEW

File: data/mods/realguns/gun.json [hunk 1]
```diff
@@ -29,6 +29,29 @@
...
+      [ "410", [ "saiga10mag_410", "saiga30mag_410" ] ],
+      [ "shot", [ "saiga10mag", "saiga30mag" ] ]
+    ]
+  },
+  {
+    "id": "saiga_410",
+    "copy-from": "saiga_12",
+    "type": "GUN",
+    "name": "Saiga-12",
+    "name_plural": "Saiga-12",
```
Comment (BorkBorkGoesTheCode) [inline data/mods/realguns/gun.json]: Why didn't you rename the Saiga to Saiga .410?

File: data/mods/realguns/gun.json [hunk 2]
```diff
@@ -29,6 +29,29 @@
...
+    "description": "A lighter version of the Saiga-12 semi-automatic shotgun, chambered in .410 bore.  Like its 12 gauge version, it uses detachable magazines.",
```
Comment (BorkBorkGoesTheCode) [inline data/mods/realguns/gun.json]: Why didn't you rename it to Saiga .410?

Comment (DangerNoodle) [inline data/mods/realguns/gun.json]: Forgot, actually.

**Comment by mugling** [comment]
We really should try and identify the source of the crash. Stack trace anyone?

**Comment by DangerNoodle** [comment]
If I knew how, that would be desirable. But fixing current instances of this might be faster, until the problem can be treated at its source.

**Comment by DangerNoodle** [comment]
Wait. @mugling, you said thar copy-from has been reverted? This mod relies heavily on self-referenced copy-from to function.

Please clarify. Because either a lot of content will be broken, or your closure of https://github.com/CleverRaven/Cataclysm-DDA/pull/19247 was unwarranted.

**Comment by mugling** [comment]
`copy-from` still works for items. The reverted PR was #19137

**Comment by mugling** [comment]
I stack traced this and  #19479 prevents the segfault. What ammo conversions are actually sensible are left to the mod maintainer.

**Comment by DangerNoodle** [comment]
Very well then. I suppose this is no longer purely needed, but implementing the Saiga-410 may still be desirable if others wish it.

**Comment by Coolthulhu** [comment]
I'll go with @mugling's PR here because it doesn't involve balancing changes.
You can still push the json changes separately, if they're desirable by themselves.

**Comment by mugling** [comment]
Agreed this PR may still be wanted - the reason for the alternative was that JSON data should **never** be able to crash the game. @DangerNoodle can you either close or confirm you want to proceed?

**Comment by DangerNoodle** [comment]
I am fine with proceeding, and I am glad that the root of the problem was sorted out so quickly.

File: data/json/items/gun/22.json [hunk 3]
```diff
@@ -23,6 +23,7 @@
     "valid_mod_locations": [ [ "accessories", 3 ], [ "barrel", 1 ], [ "bore", 1 ], [ "grip", 1 ], [ "mechanism", 4 ], [ "muzzle", 1 ], [ "rail", 1 ], [ "sights", 1 ], [ "stock", 1 ], [ "underbarrel", 1 ] ],
     "magazines": [
       [ "22", [ "a180mag" ] ],
+      [ "46", [ "hk46bigmag" ] ],
```
Comment (mugling) [inline data/json/items/gun/22.json]: Magazine is very much the wrong shape. It's not necessary that every possible conversion is possible and perhaps even desirable that some are not.

Comment (DangerNoodle) [inline data/json/items/gun/22.json]: Can remove it now that it is no longer needed.

File: data/json/items/gun/22.json [hunk 4]
```diff
@@ -179,7 +180,11 @@
     "durability": 8,
     "valid_mod_locations": [ [ "accessories", 2 ], [ "barrel", 1 ], [ "bore", 1 ], [ "grip", 1 ], [ "mechanism", 4 ], [ "muzzle", 1 ], [ "rail", 1 ], [ "sights", 1 ], [ "stock", 1 ], [ "underbarrel", 1 ] ],
     "magazine_well": 1,
-    "magazines": [ [ "22", [ "mosquitomag" ] ] ]
+    "magazines": [
```
Comment (DangerNoodle) [inline data/json/items/gun/22.json]: Can remove that as well in a moment, will be out for the moment.

File: data/json/items/gun/9mm.json [hunk 5]
```diff
@@ -62,6 +62,7 @@
       [ "9mm", [ "calicomag" ] ],
       [ "22", [ "a180mag" ] ],
       [ "45", [ "thompson_drum" ] ],
+      [ "46", [ "hk46bigmag" ] ],
```
Comment (mugling) [inline data/json/items/gun/9mm.json]: Same here, wrong shape magazine (if your bothered about reality). The balance concern is that `46` is already better than `57` and limiting conversions of the former can help to offset that

File: data/mods/realguns/magazine.json [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    "weight": 180,
+    "ammo_type": "410"
+  },
+  {
+    "id": "saiga30mag_410",
+    "copy-from": "saiga30mag",
+    "type": "MAGAZINE",
+    "name": "Saiga-410 drum magazine",
+    "description": "A removable plastic magazine for the Saiga-410 shotgun.  Holds 30 rounds of .410 bore shotshells.",
+    "weight": 400,
```
Comment (mugling) [inline data/mods/realguns/magazine.json]: You could do `relative` and `proportional` here but it's not essential

File: data/mods/realguns/gunmod.json [hunk 7]
```diff
@@ -37,7 +37,7 @@
     "color": "green",
     "location": "bore",
     "mod_targets": [ "rifle", "shotgun" ],
-    "acceptable_ammo": [ "762", "223", "308", "shot" ],
+    "acceptable_ammo": [ "shot" ],
```
Comment (DangerNoodle) [inline data/mods/realguns/gunmod.json]: Now that the crash issue is gone, I suppose that means I can implement the old conversion options without requiring I add dozens of weapon overrides, so I can undo this change if it is desirable.

File: data/json/items/gun/22.json (hunk 4 above)
Comment (DangerNoodle) [inline data/json/items/gun/22.json]: Very well then, removing the added magazine options and thus restoring this to how it was originally.

**Comment by DangerNoodle** [comment]
There we go, apologies for the delay.

**Comment by mugling** [comment]
JSON load errors
```
Checking mod Extended Realistic Guns [realguns]
Error loading data: data/mods//realguns/magazine.json: line 20:4: separator not strictly allowed at end of array
```

**Comment by DangerNoodle** [comment]
Ah, sorry about that. Had been in a hurry previously, I suspect.

**Comment by DangerNoodle** [comment]
Linting and load-testing now.

**Comment by mugling** [comment]
> Linting and load-testing now.

You can do `--check-mods realguns`

**Comment by DangerNoodle** [comment]
I still have yet to figure out how that works, nor what program is needed for that.

**Pull request merged**